### DISCUSSION TITLE
Builtin function in an "existing package"

### DISCUSSION THREAD

**Discussion by ricleal-fugue** [open-policy-agent/community#147] [Community]
Hi all,

So I have `rego` package called `pack` where I have few functions.

I created a builtin function as if it was in this package:
```go
var resourcesBuiltin = &ast.Builtin{
	Name:     "pack.resources",
	Decl:     types.NewFunction(types.Args(types.S), types.A),
	Relation: true,
}
```

I have a few modules that use this builtin function.
When I compile these modules with the builtin created:
```go
	compiler := ast.NewCompiler().WithBuiltins(map[string]*ast.Builtin{
		resourcesBuiltin.Name: resourcesBuiltin,
	})
	compiler.Compile(mods)
	if compiler.Failed() {
		return nil, compiler.Errors
	}
```
The compiler fails with the message:
```
Code:"rego_type_error"
Message:"undefined function data.pack.resources"
```

Is this normal? The builtins have to be declared in a new package?
Thanks

**Comment by anderseknert** [comment]
Built-in functions aren't bound to a package. I'm a little surprised by the error message though, as I wouldn't have expected the name to be treated as any other (i.e. without consideration for the package it was in), and this seems to be the case for other built-ins. This works fine, for example:

```rego
package object

foo := object.get(input, "bar", null)
```

How are you referencing that function in your policy?

**Reply by ricleal-fugue** [reply]
Thanks for the reply.

In the policy that I pass to the compiler I'm just calling it as a regular function:

```rego
package pack2.subpack2

import data.pack

(...)

# this a func implemented in `pack`
r = pack.func1("xpto")

# This is a call to my builtin
my_resources = pack.resources("cloud")

(...)
```

The 2nd call is the one that fails...

**Reply by anderseknert** [reply]
Thanks for clarifying. The behavior is the same for the provided built-ins:

```rego
package foo

import data.object

x = object.get(input, ["x"], null)
```

```shell
1 error occurred: foo.rego:5: rego_type_error: undefined function data.object.get
```

I.e. the `import data.object` will make import `object` into the module, and later references to `object` will assume that's what is asked for... I'm honestly not sure about the expected behavior here, but I'd suggest renaming your custom built-in to not conflict with the name of your imported package, alternatively you could us `as` to rename the import, like:

```rego
package foo

import data.object as obj

x = object.get(input, ["x"], null)
```

The custom built-in function is _not_ bound to a package, so IMHO it makes sense to not name it as if it was.

**Reply by ricleal-fugue** [reply]
Thanks for the reply and clarifying the issue.
I will name my built-in differently.
Thanks again.
